framework/model/module/module.go

```diff
@@ -43,8 +51,29 @@ type Module interface {
 	Kind() string
 	Config() proto.Message
 	InitScheme(scheme *runtime.Scheme) error
-	InitManager(mgr manager.Manager, env bootstrap.Environment, cbs InitCallbacks) error
+	// Init initialize the module according to the environment context.
+	// It is not recommended to execute business logic or start resident
+	// services at this stage, and it is forbidden to start resident
+	// services that require a single instance to run at this stage.
+	Init(env bootstrap.Environment) error
+	// SetupWithInitCallbacks registers init callbacks that supports
+	// concurrent execution, and notifies exit through `Context`, and
+	// the callback must be non-blocking.
+	SetupWithInitCallbacks(cbs InitCallbacks) error
+	// SetupWithManager registers `Reconciler` with `Manager` to build a `Controller`
+	SetupWithManager(mgr manager.Manager) error
+	// SetupWithLeaderElection registers callbacks that require a single
+	// instance to run.
+	// Generally, resident services that run concurrently and trigger the
+	// creation and update of resources in the cluster may involve race
+	// conditions and cause system exceptions. The startup of these services
+	// must be controlled through an election mechanism.
+	SetupWithLeaderElection(le leaderelection.LeaderCallbacks) error
 	Clone() Module
+
+	// For legcy moudle interface
+	InitManager(mgr manager.Manager, env bootstrap.Environment, cbs InitCallbacks) error
```

Comment:
too dirty. we can declare a new interface like `LegacyInitManager` and use type assertion to check for legacy modules.

```go
type LegacyInitManager interface {
  InitManager(mgr manager.Manager, env bootstrap.Environment, cbs InitCallbacks) error
}

for _, m := range modules {
  legacy, ok := m.(LegacyInitManager)
}
```

Reply:
done